csrc/optimization/alias_analysis.cpp

```diff
@@ -103,27 +42,151 @@ class AliasFinder : public OptOutConstDispatch {
   AliasAnalysisResult& analysis_;
 };
 
+// Computes `Split`'s output contiguity. Returns the outer contiguity and then
+// the inner contiguity.
+std::pair<std::optional<bool>, std::optional<bool>> splitContiguity(
+    const std::optional<bool>& contiguity) {
+  // Credits to @jacobhinkle:
+  // https://github.com/NVIDIA/Fuser/pull/1124#discussion_r1368682735
+  if (!contiguity.has_value()) {
+    return {std::nullopt, std::nullopt};
+  }
+  if (*contiguity) {
+    return {true, true};
+  } else {
+    return {true, false};
+  }
+}
+
+// Computes `Merge`'s output contiguity. Returns a pair
+// `<mergeable,contiguity>`. `mergeable` indicates whether the two IterDomains
+// can be merged without materialization. For example, there's no way to merge
+// `outer=f,inner=t` while keeping the output as an alias, because a dimension
+// can only have one stride. `contiguity` is the contiguity of the merged output
+// IterDomain.
+//
+// Credits to @jacobhinkle:
+// https://github.com/NVIDIA/Fuser/pull/1124#discussion_r1368682735
+std::pair<bool, std::optional<bool>> mergeContiguity(
+    const IterDomain* outer_id,
+    const std::optional<bool>& outer_contiguity,
+    const IterDomain* inner_id,
+    const std::optional<bool>& inner_contiguity) {
+  // Statuses `b` and `e` are represented in the IR with isBroadcast() and
+  // hasExpandedExtent(). Status `C` means stops propagating because we know we
+  // can't alias at that point.
+  //
+  // o\i | t  f  b  e
+  // ----+-----------
+  //  t  | t  f  t  C
+  //  f  | C  C  f  C
+  //  b  | t  f  b  e
+  //  e  | C  C  e  e
+  if (!outer_contiguity.has_value() && !outer_id->hasExpandedExtent()) {
+    return {true, inner_contiguity};
+  }
+  if (!inner_contiguity.has_value() && !inner_id->hasExpandedExtent()) {
+    return {true, outer_contiguity};
+  }
+
+  // o\i | t  f  b  e
+  // ----+-----------
+  //  t  | t  f     C
+  //  f  | C  C     C
+  //  b  |
+  //  e  | C  C     e
+  if (outer_id->hasExpandedExtent() && inner_id->hasExpandedExtent()) {
+    return {true, std::nullopt};
+  }
+  if (outer_id->hasExpandedExtent() || inner_id->hasExpandedExtent()) {
+    return {false, std::nullopt};
+  }
+
+  // o\i | t  f  b  e
+  // ----+-----------
+  //  t  | t  f
+  //  f  | C  C
+  //  b  |
+  //  e  |
+  if (*outer_contiguity) {
+    return {true, inner_contiguity};
+  }
+  return {false, std::nullopt};
+}
+
 void AliasFinder::handle(const ViewOp* view) {
   TensorView* in = view->in();
   TensorView* out = view->out();
 
+  const std::vector<IterDomain*>& in_rfactor = in->getMaybeRFactorDomain();
+  const std::vector<IterDomain*>& out_root = out->getRootDomain();
+  const std::vector<IterDomain*>& out_rfactor = out->getMaybeRFactorDomain();
+
```

Comment:
I think we forgot to check the AllocationDomain on `out` here.

Reply:
Good catch -- we should at least be consistent. We chatted about this offline. I have been on the fence whether AliasAnalysis or MarkAliasPass should respect existing allocation domains. For implementation convenience, I'm leaning towards letting the user resolve incompatibility. I added a comment around `findAliases` and adjusted some tests to reflect that.